src/main/network/packet.rs

```
@@ -421,364 +831,763 @@ impl PacketDisplay for *const c::Packet {
+        if let Some(scale) = tcp.header.window_scale {
+            c_hdr.windowScale = scale;
+            c_hdr.windowScaleSet = true;
+        }
+        c_hdr.timestampValue = to_legacy_timestamp(tcp.header.timestamp);
+        c_hdr.timestampEcho = to_legacy_timestamp(tcp.header.timestamp_echo);
+
+        c_hdr
     }
 
     #[no_mangle]
-    pub extern "C-unwind" fn packet_getDestinationIP(packet: *const c::Packet) -> libc::in_addr_t {
-        unsafe { c::legacypacket_getDestinationIP(packet) }
+    pub extern "C-unwind" fn packet_setPayloadWithMemoryManager(
+        packet_ptr: *mut Packet,
+        src: UntypedForeignPtr,
+        src_len: u64,
+        mem: *const MemoryManager,
+    ) {
+        // Read data from the managed process into the packet's payload buffer.
+        let packet = PacketRc::borrow_raw_mut(packet_ptr);
+        let mem = unsafe { mem.as_ref() }.unwrap();
+
+        let Data::LegacyTcp(tcp) = &packet.data else {
+            unimplemented!()
+        };
+
+        let len = usize::try_from(src_len).unwrap_or(usize::MAX);
+        let src = ForeignArrayPtr::new(src.cast::<u8>(), len);
+
+        // We want the dst buf on the heap so we don't have to copy it later.
+        let mut dst = vec![0u8; len].into_boxed_slice();
+
+        log::trace!(
+            "Requested to read payload of len {len} from the managed process into the packet's \
+            payload buffer",
+        );
+
+        // Copy from the managed process directly into the heap buffer. We want this to be the only
+        // copy of the payload that occurs until the receiver host later copies it into their
+        // managed process.
+        if let Err(e) = mem.copy_from_ptr(&mut dst[..], src) {
+            log::warn!(
+                "Couldn't read managed process memory at {:?} into packet payload at {:?}: {:?}",
+                src,
+                dst,
+                e
+            );
+            return;
+        }
```

If copying data into the payload fails, I think we should panic rather than just show a warning. Otherwise the packet data will be corrupt (not what the application wrote).